### PR TITLE
Search/Hotkeys: handle `/` properly when inside INPUT tags

### LINKED ISSUE #160: Search: focus input when the modal is opened
When the modal is opened (either by pressing `/` or by clicking on the search bar from the flyout), the focus isn't transferred to the input box, you need to manually focus the box to star typing.

this can be tested in https://blog.pypi.org/ for example

### LINKED ISSUE #165: Search: Glitchy behavior of search-as-you-type
## Details

* Read the Docs project URL: https://readthedocs.org/projects/kedro/

## Expected Result

Search-as-you-type behaves normally.

## Actual Result

Screen Recording 2023-10-18 at 13 36 10

This is on both Firefox 118.0.2 and Safari on macOS Ventura. We just enabled the Addons.

cc @stichbury

### PR DESCRIPTION
**Pull request by humitos**
- do not show search modal when inside `input` HTML tags
- prevent keypress to propagate when captured by the hotkeys addons
- don't handle `/` when inside `readthedocs-search` HTML tag
- focus input field from the modal immediately after pressing `/`

Closes #165
Closes #160

### CODE REVIEW

**Comment by agjohnson** [comment]
I'm seeing the following runtime errors on the main run dev page (localhost:8000/):

```
Cannot read properties of null (reading 'focus')
TypeError: Cannot read properties of null (reading 'focus')
    at SearchElement.updated (http://localhost:8000/readthedocs-addons.js:1726:15)
    at SearchElement._$didUpdate (http://localhost:8000/readthedocs-addons.js:25170:14)
    at SearchElement.performUpdate (http://localhost:8000/readthedocs-addons.js:25136:18)
    at SearchElement.scheduleUpdate (http://localhost:8000/readthedocs-addons.js:25055:21)
    at SearchElement.__enqueueUpdate (http://localhost:8000/readthedocs-addons.js:25028:29)
```

The search modal doesn't work on the Sphinx example either. `/` starts typing in the search input box.

It does come up on the Pelican example however, though the modal doesn't close with `esc`.

This is with an fresh browser environment (chromium), so no addons should be interfering.

**Review by agjohnson — COMMENTED** [review]
The code here looks correct, but the problem with `focus()` is probably a matter of triggering a DOM event on a node that doesn't exist yet.

I don't think this change would negatively impact accessibility, but at some point we should probably return and be testing this with accessibility tooling more.

**Comment by humitos** [comment]
> The search modal doesn't work on the Sphinx example either. `/` starts typing in the search input box.

This is because Sphinx captures the `/` key as well. We need to document how to disable it on Sphinx for users wanting to show our own search modal when using the hotkey addons. I think it's gonna be something like:

```python
# docs/conf.py

html_theme_options = {
  "enable_search_shortcuts": False,
}
```

Reference: https://www.sphinx-doc.org/en/master/usage/theming.html#builtin-themes

**Comment by humitos** [comment]
> It does come up on the Pelican example however, though the modal doesn't close with `esc`.

IIRC, we haven't implemented `Esc` to close the modal yet. It's only on the modal's bottom as documentation, but it's not on the code 🙃 . Opened https://github.com/readthedocs/addons/issues/172 to track this work.

File: src/search.js [hunk 1]
```diff
@@ -286,7 +286,7 @@ export class SearchElement extends LitElement {
   updated(changedProperties) {
     // https://lit.dev/docs/components/shadow-dom/
     const input = this.shadowRoot.querySelector("input[type=search]");
-    if (input === undefined) {
+    if (input !== undefined) {
       input.focus();
```
[the remark this answers is not in the document]
Comment (humitos) [reply src/search.js:290]: Hrm... Interesting. We talked about this already in #115 and I thought that `updated()` was the correct method to use here. I'll see if I find another Lit method that works better for our use case here.

[the remark this answers is not in the document]
Comment (humitos) [reply src/search.js:290]: I tried `firstUpdated` but that doesn't work.

The problem was that `input` was not `undefined`, but it was `null`. So, I'm checking for `!== undefined and !== null` now. That makes the code to work fine.

**Comment by agjohnson** [comment]
> I suppose we will have these types of hints/snippets for different doctools while we are finding this incompatibilities.

Yeah, I figure the same. Especially where we have both user and theme hotkey customizations to consider. These hurdles are some of the reasons I've been very hesitant to implement hotkey features on any of our product side. I think there is plenty of room to use this pattern for testing purposes though, and wouldn't change our existing patterns. But I eventually see us implementing UI instead of hotkeys as the primary method for invoking these features, leaving keyboard access as a secondary or optional method.

File: src/search.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (agjohnson) [reply src/search.js:290]: Yeah, that makes sense, though I can't say exactly what would be defining it as `null` either. I'd agree that checking for `null` probably solves this issue though.

**Review by agjohnson — APPROVED** [review]

**Comment by humitos** [comment]
> leaving keyboard access as a secondary or optional method.

Yes, I suppose we will be making this opt-in while explaining the potential incompatibilities and linking to our documentation where we explain the known-issues and how to disable conflicting keys.

I think it's _impossible_ to make hotkeys to always work fine on "any HTML the user can create" 🙃 . In any case, I'm happy we have this addon at this moment and we can keep experimenting with it to discover these things.

I may change my mind in the future if I find a way to "disable all the hotkeys for any other JS and only listen to mines", but I think that's not possible anyways 😄

**Pull request merged**